subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -7,22 +7,250 @@
+
+static int modem_fw_update(void)
+{
+	int rc;
+
+	LOG_INF("Scheduling modem firmware update at next boot");
+	rc = nrf_setsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_APPLY,
+			    NULL, 0);
+	if (rc < 0) {
+		LOG_INF("Failed to schedule update, errno %d", errno);
+	}
+
+	return 0;
+}
+
+static int modem_fw_delete(void)
+{
+	int rc;
+
+	LOG_INF("Deleting firmware image, please wait...\n");
+	rc = nrf_setsockopt(modem_dfu_socket_fd, NRF_SOL_DFU,
+			    NRF_SO_DFU_BACKUP_DELETE, NULL, 0);
+	if (rc < 0) {
+		LOG_ERR("Failed to delete backup, errno %d\n", errno);
+		return 0;
+	}
+
+	k_delayed_work_submit_to_queue(&modem_fw_delete_offload_work_q,
+				       &modem_fw_delete_dwork, K_SECONDS(1));
+
+	return 0;
+}
+
+/**@brief Initialize DFU socket. */
+static int modem_dfu_socket_init(void)
+{
+	int rc;
+	nrf_socklen_t len;
+	nrf_dfu_fw_version_t version;
+	nrf_dfu_resources_t resource;
+	nrf_dfu_fw_offset_t off;
+
+	/* Create a socket for firmware update */
+	modem_dfu_socket_fd = nrf_socket(NRF_AF_LOCAL, NRF_SOCK_STREAM,
+				     NRF_PROTO_DFU);
+	if (modem_dfu_socket_fd < 0) {
+		LOG_ERR("Failed to open Modem DFU socket.");
+		return -1;
+	}
+
+	LOG_INF("Modem DFU Socket created");
+
+	len = sizeof(version);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU,
+			    NRF_SO_DFU_FW_VERSION, &version, &len);
+	if (rc < 0) {
+		LOG_ERR("Firmware version request failed, errno %d", errno);
+		return -1;
+	}
+
+	*((u8_t *)version + sizeof(version) - 1) = '\0';
+	LOG_DBG("Modem firmware version: %s", log_strdup(version));
+
+	len = sizeof(resource);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU,
+			    NRF_SO_DFU_RESOURCES, &resource, &len);
+	if (rc < 0) {
+		LOG_ERR("Resource request failed, errno %d", errno);
+		return -1;
+	}
+
+	LOG_DBG("Flash memory available: %u", resource);
+
+	len = sizeof(off);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			    &off, &len);
+	if (rc < 0) {
+		LOG_ERR("Offset request failed, errno %d", errno);
+		return -1;
+	}
+
+	LOG_DBG("Offset retrieved: %u", off);
+
+	return rc;
+}
+
+static nrf_dfu_err_t modem_error_get(void)
+{
+	int rc;
+	nrf_dfu_err_t err;
+	nrf_socklen_t len;
+
+	len = sizeof(err);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_ERROR,
+			    &err, &len);
+	if (rc) {
+		LOG_ERR("Unable to fetch modem error, errno %d", errno);
+	}
+
+	return err;
+}
+
+static int modem_fw_write(u8_t *buf, size_t len)
+{
+	int sent = send(modem_dfu_socket_fd, buf, len, 0);
+
+	if (sent < 0) {
+		int modem_error = modem_error_get();
+
+		LOG_ERR("Modem refused fragment, errno %d, dfu err %d", errno,
+			modem_error);
+		if (modem_error == DFU_ERASE_PENDING) {
+			return -EAGAIN;
+		}
+		return -1;
+	}
+	return 0;
+}
+
+static void modem_fw_delete_task(struct k_work *w)
+{
+	int rc;
+	u32_t off;
+	nrf_socklen_t len;
+
+	len = sizeof(off);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
```

Comment:
From what I gathered from looking at @lemrey example and running it you basically query the modem through getsocketopt with offset until the modem returns with success instead of `DFU_ERASE_PENDING`.  Hence the delayed work re-scheduling.

Reply:
There is no callback mechanism from the DFU interface and the application, hence the application has to poll the offset to determine whether the erase has taken place or not. Once the erase operation has started the modem will return a specific error indicating that the poll operation has not been executed, and an extended error code can be fetch and indicates that the reason is that an erase operation is pending.

So in practice the code is correct, it could additonally check for these specific errors; I'll admit the API can use some improvement.

Reply:
I understand that, my point is that this piece of code could use some explanation in the comment of why is this done this way. Otherwise, someone who revisits this code in the future might introduce some regression because of a lack of understanding of this approach.

Reply:
I'll add a comment explaining the functionality.

Reply:
Fixed